### PR TITLE
[MGDAPI-4259] update 3scale portal access alerts to account for custom domains

### PR DESCRIPTION
**Pull request by cecobask**
# Issue link
https://issues.redhat.com/browse/MGDAPI-4259

# What

# Verification steps
## Upgrade scenario
1. Provision/borrow an OSD cluster
2. `make cluster/prepare` from the master branch
3. Create a CatalogSource CR and install RHOAM from it:
```
oc apply -f - <<EOF
---
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhmi-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/tdimov0/managed-api-service-index:1.24.0
EOF
```
4. Wait for the RHOAM installation to complete
5. Modify the CatalogSource to point at `quay.io/tdimov0/managed-api-service-index:1.24.1`:
```
oc patch catalogsource rhmi-operators -n openshift-marketplace --type=json -p='[{"op": "replace", "path": /spec/image, "value": "quay.io/tdimov0/managed-api-service-index:1.24.1"}]'
```
6. Initiate RHOAM upgrade
7. Verify that the upgrade completes and RHMI CR is set to version `1.24.1`

## Fresh install scenario
1. Delete the RHMI CR and wait for successful deletion
2. Make sure the CatalogSource points at `quay.io/tdimov0/managed-api-service-index:1.24.1`:
```
oc patch catalogsource rhmi-operators -n openshift-marketplace --type=json -p='[{"op": "replace", "path": /spec/image, "value": "quay.io/tdimov0/managed-api-service-index:1.24.1"}]'
```
3. Follow [this guide](https://github.com/cecobask/integreatly-operator/blob/MGDAPI-4259/guides/custom_domain.md) to create the required resources for custom domain
4. Wait for the RHOAM installation to complete
5. Verify the `CustomDomainThreeScaleAdminUI`, `CustomDomainThreeScaleDeveloperUI`, `CustomDomainThreeScaleSystemAdminUI` did not fire during installation or after that. If the query `rhoam_custom_domain` returns a graph similar to this with labels `active="true", system_developer="true", system_master="true", system_provider="true"`, then the verification is successful
image
6. To trigger the alerts, you can scale down the 3scale operator and the system-app:
```
oc patch deployment threescale-operator-controller-manager-v2 -n redhat-rhoam-3scale-operator --type=json -p='[{"op": "replace", "path": /spec/replicas, "value": 0}]'
```
```
oc patch deploymentconfig system-app -n redhat-rhoam-3scale --type=json -p='[{"op": "replace", "path": /spec/replicas, "value": 0}]'
```
7. After 15 minutes, verify the 3 alerts from the above step fired
8. Scale the 3scale operator from 0 back to 1 replica:
```
oc patch deployment threescale-operator-controller-manager-v2 -n redhat-rhoam-3scale-operator --type=json -p='[{"op": "replace", "path": /spec/replicas, "value": 1}]'
```
9. After the operator restores the system-app, verify the 3 alerts are not firing anymore

### CODE REVIEW

File: pkg/products/threescale/reconciler.go [hunk 1]
```diff
@@ -441,6 +467,34 @@ func (r *Reconciler) Reconcile(ctx context.Context, installation *integreatlyv1a
...
+		ingressRouterService, err := customDomain.GetIngressRouterService(ctx, serverClient)
+		if err != nil {
+			errorMessage := "failed to retrieve ingress router service"
+			r.log.Error(errorMessage, err)
+			phase = integreatlyv1alpha1.PhaseFailed
+			events.HandleError(r.recorder, installation, phase, errorMessage, err)
+			return phase, fmt.Errorf("%s: %v", errorMessage, err)
+		}
+
+		ips, err := customDomain.GetIngressRouterIPs(ingressRouterService.Status.LoadBalancer.Ingress[0].Hostname)
```
Comment (briangallagher) [inline pkg/products/threescale/reconciler.go:480]: do we need nil checks here?

File: pkg/products/rhssocommon/reconciler.go [hunk 2]
```diff
@@ -328,7 +328,7 @@ func (r *Reconciler) SetupOpenshiftIDP(ctx context.Context, serverClient k8sclie
 			FirstBrokerLoginFlowAlias: "first broker login",
 			Config: map[string]string{
 				"hideOnLoginPage": "",
-				"baseUrl":         "https://" + strings.Replace(r.Installation.Spec.RoutingSubdomain, "apps", "api", 1) + ":6443",
+				"baseUrl":         "https://" + strings.Replace(r.Installation.Spec.MasterURL, "console-openshift-console.apps", "api", 1) + ":6443",
```
[the remark this answers is not in the document]
Comment (cecobask) [reply pkg/products/rhssocommon/reconciler.go:331]: This was tested on both custom domain and regular RHOAM installations. Shouldn't be a problem, but if the format changes in the future, then this code has to be updated too

[the remark this answers is not in the document]
Comment (adam-cattermole) [reply pkg/products/rhssocommon/reconciler.go:331]: You're right Brian that the MasterURL format can be changed and I wondered about this too - the user can go and modify the route to the openshift console. Do we know if it's common for this route to be modified?

An alternative solution is adding another field to the RHMI spec to store the cluster API URL which would be safer

[the remark this answers is not in the document]
Comment (briangallagher) [reply pkg/products/rhssocommon/reconciler.go:331]: I'm not sure how common it is but think the alternative solution is the way to go here. As you said, safer.

File: controllers/rhmi/bootstrapReconciler_test.go [hunk 3]
```diff
@@ -800,3 +802,229 @@ func TestReconciler_reconcileAddonManagedApiServiceParameters(t *testing.T) {
...
+								Namespace: "test",
+							},
+							Data: map[string][]byte{
+								"custom-domain_domain": []byte("{ } < > | ` ^"),
+							},
+						})
+				},
+			},
+			want:    integreatlyv1alpha1.PhaseCompleted,
+			wantErr: false,
```
Comment (Patryk-Stefanski) [inline controllers/rhmi/bootstrapReconciler_test.go:917]: If the domain name is invalid, should it return an error?

File: pkg/products/rhssocommon/reconciler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cecobask) [reply pkg/products/rhssocommon/reconciler.go:331]: Let's hold merging this PR until this one is merged: https://github.com/integr8ly/integreatly-operator/pull/2773. I'll then rebase my one, as Adam suggested.

File: controllers/rhmi/bootstrapReconciler_test.go (hunk 3 above)
Comment (Boomatang) [reply controllers/rhmi/bootstrapReconciler_test.go:917]: To add some more context to this. We do not block during the bootstrap reconcile to allow time for our monitoring stack to be installed. Once our monitoring stack is installed we can send alerts to the customers. If we where to block in the bootstrap stage these alerts would have to go through SRE-P which makes getting alerts to the customers complicated. We do block at the 3scale reconcile if there is any issue found.

**Comment by Boomatang** [comment]
I have done the verification steps and am happy with the results. This PR will need to be rebased once #2773 is merge to master. Waiting till that happens before given the lgtm

**Comment by Boomatang** [comment]
/lgtm

This all looks good. The verification steps I carried out include triggering the alerts on upgrades, fresh standard installations and custom domain installation. Some of the latest changes requests happen on https://github.com/Boomatang/integreatly-operator/pull/3 that PR was against the branch this PR is based off. It made the reviewing of changes easier

**Pull request merged**